util/elf-writer.cc

```diff
@@ -0,0 +1,280 @@
+// limitations under the License.
+
+#include "util/elf-writer.h"
+
+#include <elf.h>
+#include <string.h>
+#include <stdio.h>
+
+// Offset for global symbol indices.
+static const int GLOBAL_SYMBOL_OFFSET = 1 << 20;
+
+Elf::Elf() {
+  // Initialize header.
+  memset(&ehdr_, 0, sizeof(ehdr_));
+  ehdr_.e_ident[EI_MAG0] = ELFMAG0;
+  ehdr_.e_ident[EI_MAG1] = ELFMAG1;
+  ehdr_.e_ident[EI_MAG2] = ELFMAG2;
+  ehdr_.e_ident[EI_MAG3] = ELFMAG3;
+  ehdr_.e_ident[EI_CLASS] = ELFCLASS64;
+  ehdr_.e_ident[EI_DATA] = ELFDATA2LSB;
+  ehdr_.e_ident[EI_VERSION] = EV_CURRENT;
+  ehdr_.e_ident[EI_OSABI] = ELFOSABI_SYSV;
+  ehdr_.e_type = ET_REL;
+  ehdr_.e_machine = EM_X86_64;
+  ehdr_.e_version = EV_CURRENT;
+  ehdr_.e_ehsize = sizeof(ehdr_);
+  ehdr_.e_shoff = sizeof(ehdr_);
+  ehdr_.e_shentsize = sizeof(Elf64_Shdr);
+
+  // Add null symbol as first symbol.
+  AddSymbol("", false);
+
+  // Add null section as first section.
+  AddSection("", SHT_NULL);
+
+  // Add section for symbol table.
+  symtab_ = AddSection(".symtab", SHT_SYMTAB);
+}
+
+Elf::~Elf() {
+  for (auto *section : sections_) delete section;
+  for (auto *symbol : local_symbols_) delete symbol;
+  for (auto *symbol : global_symbols_) delete symbol;
+}
+
+Elf::Section *Elf::AddSection(const char *name, Elf64_Word type) {
+  // Allocate new section.
+  Section *section = new Section(sections_.size());
+  sections_.push_back(section);
+
+  // Set section name.
+  int namelen = strlen(name);
+  section->hdr.sh_name = section_names_.size();
+  section_names_.append(name, namelen + 1);
+
+  // Set section type.
+  section->hdr.sh_type = type;
+  if (type != SHT_NULL) section->hdr.sh_addralign = 1;
+
+  // Add symbol for section.
+  if (type == SHT_PROGBITS) {
+    section->symidx = AddSymbol("", section, STB_LOCAL, STT_SECTION)->index;
+  }
+
+  return section;
+}
+
+Elf::Symbol *Elf::AddSymbol(const char *name, bool global) {
+  // Allocate new symbol.
+  Symbol *symbol;
+  if (global) {
+    symbol = new Symbol(global_symbols_.size() + GLOBAL_SYMBOL_OFFSET);
+    global_symbols_.push_back(symbol);
+  } else {
+    symbol = new Symbol(local_symbols_.size());
+    local_symbols_.push_back(symbol);
+  }
+
+  // Set symbol name.
+  int namelen = strlen(name);
+  if (namelen == 0 && !symbol_names_.empty()) {
+    symbol->sym.st_name = 0;
+  } else {
+    symbol->sym.st_name = symbol_names_.size();
+    symbol_names_.append(name, namelen + 1);
+  }
+
+  return symbol;
+}
+
+Elf::Symbol *Elf::AddSymbol(const char *name, Section *section,
+                            int bind, int type, int size, int value) {
+  Symbol *symbol = AddSymbol(name, bind == STB_GLOBAL);
+  symbol->sym.st_info = ELF64_ST_INFO(bind, type);
+  if (section != nullptr) {
+    symbol->sym.st_shndx = section->index;
+  }
+  symbol->sym.st_size = size;
+  symbol->sym.st_value = value;
+  return symbol;
+}
+
+void Elf::Update() {
+  // Build symbol table. Local symbols must come before non-local symbols in the
```

Comment:
I have added a comment to both Elf::Update() and Elf::Buffer::Update(). I cannot use CHECK macros in this code because it is also used during the build process for embedding data in the binary, so to avoid any circular dependencies, the elf-writer module does not depend on any other SLING code. This is also the reason that I use fopen() instead of File::Open().